### PR TITLE
Add possibility to record branch information in an issue

### PR DESCRIPTION
**Pull request by gzsombor**
This patch add a new field to the 'issue' entity, named 'branch' with the necessary UI to specify which branch this issue is referring. 
 Searching, filtering and editing is not yet implemented, so the branch can only specified when creating the issue

### CODE REVIEW

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -531,6 +531,7 @@ issues.new.closed_milestone = Closed Milestones
 issues.new.assignee = Assignee
 issues.new.clear_assignee = Clear assignee
 issues.new.no_assignee = No assignee
+issues.noBranch = No Branch Specified
```
Comment (andreynering) [inline options/locale/locale_en-US.ini]: `noBranch` -> `no_branch`

File: templates/repo/issue/new_form.tmpl [hunk 2]
```diff
@@ -32,6 +32,27 @@
 
 	<div class="four wide column">
 		<div class="ui segment metas">
+			<input id="branch_selector" name="branch" type="hidden" value="{{.branch}}">
+			<div class="ui floating filter select-branch dropdown" data-no-results="{{.i18n.Tr "repo.pulls.no_results"}}">
+				<div class="ui basic small button">
+					<span class="text branch-name">{{if .branch}}{{.branch}}{{else}}{{.i18n.Tr "repo.issues.noBranch"}}{{end}}</span>
```
Comment (andreynering) [inline templates/repo/issue/new_form.tmpl]: `noBranch` -> `no_branch`

File: templates/repo/issue/view_content.tmpl [hunk 3]
```diff
@@ -253,6 +253,26 @@
 
 	<div class="four wide column">
 		<div class="ui segment metas">
+			<div class="ui disabled floating filter select-branch dropdown" data-no-results="{{.i18n.Tr "repo.pulls.no_results"}}">
+				<div class="ui basic small button">
+					<span class="text branch-name">{{if .Issue.Branch}}{{.Issue.Branch}}{{else}}{{.i18n.Tr "repo.issues.noBranch"}}{{end}}</span>
```
Comment (andreynering) [inline templates/repo/issue/view_content.tmpl]: `noBranch` -> `no_branch`

**Review by andreynering — COMMENTED** [review]
I'd like to hear the @go-gitea/owners opinion on whether we will accept or not this feature.

**Comment by lunny** [comment]
I think this PR is good point, but maybe also need tag or commit? Of course, put it in v1.2 is a correct decision. @andreynering

**Comment by gzsombor** [comment]
I've implemented the requested changes from noBranch to no_branch
As the branch field is just a string, theoretically it could also contain a tag or a commit, if the frontend component is changed

**Comment by bkcsoft** [comment]
@lunny better to track the branch only. GitLab also has this feature, very useful when an issue has many PRs :+1:

**Comment by tboerger** [comment]
I like this feature, but I would like to be able to also select a tag. Pretty good for bug reports for deployed applications.

**Comment by gzsombor** [comment]
Do I need to do anything with it to merge ?
I guess, it's not really hard to make this field configurable - so on a repository level, the admin can say, if it's needs this field, it should contain branch, or tag, or both. But that should go to a different commit :)

**Comment by lunny** [comment]
I will really review this after 1.1 released.

**Comment by lunny** [comment]
@gzsombor could you add tag selection also?

**Comment by lunny** [comment]
conflict

**Comment by gzsombor** [comment]
Patch is updated with tag selection

**Comment by gzsombor** [comment]
What do you think?

**Comment by lunny** [comment]
@gzsombor I will review this again.

File: models/issue.go [hunk 4]
```diff
@@ -49,6 +49,7 @@ type Issue struct {
 	IsPull          bool         `xorm:"INDEX"` // Indicates whether is a pull request or not.
 	PullRequest     *PullRequest `xorm:"-"`
 	NumComments     int
+	Branch          string
```
Comment (bkcsoft) [inline models/issue.go]: Maybe call this `Ref` instead, since we can track both Branches and Tags?

File: modules/auth/repo_form.go [hunk 5]
```diff
@@ -194,6 +194,7 @@ func (f *NewSlackHookForm) Validate(ctx *macaron.Context, errs binding.Errors) b
 type CreateIssueForm struct {
 	Title       string `binding:"Required;MaxSize(255)"`
 	LabelIDs    string `form:"label_ids"`
+	Branch      string `form:"branch"`
```
Comment (bkcsoft) [inline modules/auth/repo_form.go]: Same here, `Ref`

File: options/locale/locale_en-US.ini [hunk 6]
```diff
@@ -600,6 +600,7 @@ issues.new.closed_milestone = Closed Milestones
 issues.new.assignee = Assignee
 issues.new.clear_assignee = Clear assignee
 issues.new.no_assignee = No assignee
+issues.no_branch = No Branch Specified
```
Comment (bkcsoft) [inline options/locale/locale_en-US.ini]: `issues.no_ref = No Branch/Tag Specified`

**Review by bkcsoft — COMMENTED** [review]
Just semanital cleanup, otherwise LGTM

**Comment by gzsombor** [comment]
I've fixed the issues

**Comment by bkcsoft** [comment]
LGTM

**Comment by lafriks** [comment]
LGTM

**Comment by wxiaoguang** [comment]
After 8 years, this "branch selector" does nothing more than saving the branch/tag name into database and display it.

Are there real users who uses this feature? If no, could it be removed in new versions?

**Comment by wxiaoguang** [comment]
For anyone who still needs it, please provide feedbacks to  Prepare to remove the branch/tag selector in the issue sidebar #32744

**Comment by wxiaoguang** [comment]
No feedback, then this "branch/tag selector" will be removed in 1.23

**Pull request merged**